src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h

```diff
@@ -0,0 +1,160 @@
+/*
+ * ExpectationMaximization.h
+ * Provides implementation of Expectation Maximization algorithm 
+ * used for Gaussian Mixture Modelling.
+ *
+Copyright 2017 Michal Gallus
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+#include <random>
+#include "GaussianMixtureModel.h"
+#include "DataType.h"
+#include "ArgumentNullException.h"
+
+namespace Spectre::libGaussianMixtureModelling
+{
+    /// <summary>
+    /// Class serves as the container for all data required by
+    /// A. P. Dempster's Expectation Maximization algorithm used
+    /// for the purpose of Guassian Mixture Modelling of the
+    /// spectral data.
+    /// </summary>
+    /// <param name="InitializationRunner">Class performing Initialization step of the em algorithm.</param>
+    /// <param name="ExpectationRunner">Class performing expectation step of the em algorithm.</param>
+    /// <param name="MaximizationRunner">Class performing maximization step of the em algorithm.</param>
+    /// <param name="LogLikelihoodCalculator">Class performing log likelihood of resulting calculation.</param>
+    template<typename InitializationRunner, typename ExpectationRunner, typename MaximizationRunner, typename LogLikelihoodCalculator>
+	class ExpectationMaximization
+	{
+	public:
+        /// <summary>
+        /// Constructor initializing the class with all algorithm necessary data.
+        /// </summary>
+        /// <param name="mzArray">Array of m/z values.</param>
+        /// <param name="intensities">Set of corresponding mean intensities values.</param>
+        /// <param name="size">Size of the mzArray and itensities arrays.</param>
+        /// <param name="rngEngine">Mersenne-Twister engine to be used during initialization step.</param>
+        /// <param name="numberOfComponents">Number of Gaussian components that build up the approximation.</param>
+        /// <exception cref="ArgumentNullException">Thrown when either of mzArray or intensities pointers are null</exception>
+		ExpectationMaximization(DataType* mzArray, DataType* intensities, const unsigned size, 
+                std::mt19937_64& rngEngine, const unsigned numberOfComponents = 2)
```

Comment:
If we provide it from somewhere else, we need to assure that this object living elsewhere and being only related to EM class will not fall out of scope. If we do it that way, a reasonable approach would be to do the same to the whole rest, to remain consequent. I do believe that the current way of doing this obscures the code in a lower extend and is at least not error-prone.